kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/encrypt/EncryptionDekCache.java

```diff
@@ -114,18 +167,19 @@ private void afterCacheEviction(@Nullable CacheKey<K> cacheKey,
      * @param filterThreadExecutor The filter thread executor.
      * @return A stage that completes on the filter thread with the DEK.
      */
-    public @NonNull CompletionStage<Dek<E>> get(@NonNull EncryptionScheme<K> encryptionScheme,
-                                                @NonNull FilterThreadExecutor filterThreadExecutor) {
+    public @NonNull CompletionStage<DekContext<E>> get(@NonNull EncryptionScheme<K> encryptionScheme,
+                                                       @NonNull FilterThreadExecutor filterThreadExecutor) {
         return filterThreadExecutor.completingOnFilterThread(dekCache.get(cacheKey(encryptionScheme)));
     }
 
-    /**
-     * Discard any cached DEK for the KEK in the given {@code encryptionScheme}.
-     * This method may block if a DEK for the given {@code encryptionScheme} is in the process
-     * of being loaded.
-     * @param encryptionScheme The KEK for the DEK to discard.
-     */
-    public void invalidate(@NonNull EncryptionScheme<K> encryptionScheme) {
-        dekCache.synchronous().invalidate(cacheKey(encryptionScheme));
+    private CompletableFuture<DekContext<E>> invalidate(@NonNull CacheKey<K> key) {
+        invalidationCount.incrementAndGet();
+        return dekCache.asMap().remove(key).whenComplete((dekContext, throwable) -> {
```

Comment:
Yep I misunderstood, `invalidate` makes sense. Though I wonder if we could make use of `asMap().computeIfPresent` instead to check the identity of the current entry and do the equivalent `remove` logic if it's the expected DEK.

Reply:
Had a play with that and it works, but without a Future to wait on the threads that don't need to invalidate can quickly re-attempt and encounter the same exhausted/destroyed DEK in a tight loop, leading to non-deterministic unit tests where the retry limits are hit (even increasing it to 100 retries). Maybe we should add some small delay on re-attempt?

With production settings a small delay shouldn't be a problem, we shouldn't be rotating DEKs so often.

Reply:
here's that alternative if you want to have a look at the approach before I polish it up https://github.com/robobario/kroxylicious/commit/21e0777ffb65b4df9d6a60f53eaa868ee9849ad4

It avoids the DekContext by leaning on `computeIfPresent` on the cache, and instead of having to share a Future for co-ordination we slightly delay the re-attempts to avoid tight retry loops.

Reply:
Alternative idea.  Did you look at com.github.benmanes.caffeine.cache.LoadingCache#refresh already?  The problem we have we the existing code is that invalidate gets called by many threads, so we have a invalidation flurry, right?   So if we were to replace `invalidate()` with `#refresh` (which is documented as "deduplicating requests") would that solve the problem?

Reply:
Yep it does have that nice property, `com.github.benmanes.caffeine.cache.LoadingCache#refresh` returns:

> Returns an existing future without doing anything if another thread is currently loading the value for key.
>
> Returns:
> the future that is loading the value

So multiple threads could chain onto the refresh future. But, I think the problem is this:

> While the new value is loading the previous value (if any) will continue to be returned by get(key) unless it is evicted.

So there can be races like:
```
1. thread 1 obtains exhausted DEK A
2. thread 1 calls refresh and chains onto the future
3. thread 2 obtains exhausted DEK A
4. refresh completes, DEK B replaces DEK A in the cache, thread 1's future is completed with DEK B
4. thread 2 calls refresh and chains onto the future
5. refresh completes. DEK C replaces DEK B in the cache, thread 2's future is completed with DEK C
```

Similar to the current issue with invalidation, it may trigger additional unexpected refreshes.

Reply:
Yes, I think you are right.  Drat!